Libplanet/Store/DefaultStore.cs

```diff
@@ -246,174 +271,237 @@ public override IEnumerable<TxId> IterateStagedTransactionIds()
+                        }
+                    }
                 }
                 finally
                 {
-                    _txLock.ExitWriteLock();
+                    if (_txs.FileExists(tmpPath))
+                    {
+                        _txs.DeleteFile(tmpPath);
+                    }
                 }
             }
-            finally
-            {
-                _txLock.ExitUpgradeableReadLock();
-            }
+
+            _txCache.AddOrUpdate(tx.Id, tx);
         }
 
         /// <inheritdoc/>
         public override bool DeleteTransaction(TxId txid)
         {
-            _txLock.EnterWriteLock();
-            try
+            var path = TxPath(txid);
+            if (_txs.FileExists(path))
             {
-                return _db.FileStorage.Delete(TxFileId(txid));
-            }
-            finally
-            {
-                _txLock.ExitWriteLock();
+                _txs.DeleteFile(path);
+                _txCache.Remove(txid);
+                return true;
             }
+
+            return false;
         }
 
         /// <inheritdoc/>
         public override IEnumerable<HashDigest<SHA256>> IterateBlockHashes()
         {
-            _blockLock.EnterReadLock();
-            try
+            foreach (UPath path in _blocks.EnumerateDirectories(UPath.Root))
             {
-                return _db.FileStorage
-                    .Find("block/")
-                    .Select(file => new HashDigest<SHA256>(ByteUtil.ParseHex(file.Filename)));
-            }
-            finally
-            {
-                _blockLock.ExitReadLock();
+                string upper = path.GetName();
+                if (upper.Length != 2)
+                {
+                    continue;
+                }
+
+                foreach (UPath subPath in _blocks.EnumerateFiles(path))
+                {
+                    string lower = subPath.GetName();
+                    if (lower.Length != 62)
+                    {
+                        continue;
+                    }
+
+                    string name = upper + lower;
+                    HashDigest<SHA256> blockHash;
+                    try
+                    {
+                        blockHash = new HashDigest<SHA256>(ByteUtil.ParseHex(name));
+                    }
+                    catch (Exception)
+                    {
+                        continue;
+                    }
+
+                    yield return blockHash;
+                }
             }
         }
 
         /// <inheritdoc/>
         public override void PutBlock<T>(Block<T> block)
         {
-            // checks whether a block already exists or not.
-            if (!(_db.FileStorage.FindById(BlockFileId(block.Hash)) is null))
+            if (_blockCache.ContainsKey(block.Hash))
             {
                 return;
             }
 
-            _blockLock.EnterWriteLock();
-            try
+            foreach (Transaction<T> tx in block.Transactions)
```

Comment:
How about checking the block not only from the cache but also from the file system?

Reply:
Fixed: c8c6bddda51893d02df6c130c02e51f7ca7ca3bd.